drivers/VirtualWatchdog.h

```diff
@@ -0,0 +1,146 @@
+/*
+ * Copyright (c) 2019 Arm Limited and affiliates.
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef MBED_VIRTUAL_WATCHDOG_H
+#define MBED_VIRTUAL_WATCHDOG_H
+
+#ifdef DEVICE_WATCHDOG
+
+#include <cstdio>
+#include "platform/mbed_error.h"
+#include "platform/mbed_critical.h"
+#include "platform/mbed_power_mgmt.h"
+#include "platform/mbed_assert.h"
+#include "platform/SingletonPtr.h"
+#include "drivers/Watchdog.h"
+
+#if DEVICE_LPTICKER
+#include "drivers/LowPowerTicker.h"
+#else
+#include "drivers/Ticker.h"
+#endif
+namespace mbed {
+
+/** \addtogroup drivers */
+
+/** VirtualWatchdog should be used for applications where you use multiple services requiring watchdog functionality.
+ *   Use `Watchdog` driver for simple uses cases like bootloader. Do not use both drivers at the same time - VirtualWatchdog will error
+ *   if it is not the only owner of the watchdog.
+ *
+ *  A system timer that will reset the system in the case of system failures or
+ *  malfunctions.
+ *
+ * Example:
+ * @code
+ *
+ * Example:
+ * @code
+ *
+ * VirtualWatchdog virtual_watchdog(300,"Software Watchdog");
+ * virtual_watchdog.start();
+ *
+ * while (true) {
+ *    virtual_watchdog.kick();
+ *
+ *    // Application code
+ * }
+ * @endcode
+ * @ingroup drivers
+ */
+class VirtualWatchdog {
+public:
+
+    /** Constructor configured with timeout and name for this software watchdog instance.
+     *
+     *  @param timeout Timer timeout
+     *  @param str The name for this watchdog timer
+     *  @param watchdog_hw_timeout The watchdog driver timeout - can be set only once (as soon as watchdog runs, cannot be reconfigured)
+     *
+     */
+    VirtualWatchdog(uint32_t timeout = 1000, const char *const str = NULL, uint32_t watchdog_hw_timeout = Watchdog::watchdog_timeout);
```

Comment:
Why would user care about the HW timeout when using VirtualWatchdog? We should handle the complexity ourselves. I still think we shouldn't have a config option to set default timeout, for Watchdog it should be just a default parameter, but for VirtualWatchdog we should figure out the right value.

Reply:
+1 for config value removal, that one is done , will be pushed. 

Updated, timeout is set on the first run based on timeout value.